### DISCUSSION TITLE
Chapter3. 함수 인수 질문

### DISCUSSION THREAD

**Discussion by CODe5753** [Well-Founded-Confidence/CleanCode#6] [질의응답] [answered]
50페이지 `함수 인수` 파트에 이런 내용이 있습니다.
> includeSetupPageInto(new PageContent)보다 includeSetupPage()가 _이해하기 더 쉽다._

뒤에 추가적인 내용으로 설명이 되어 있으나, 왜 `includeSetupPageInto(new PageContent)`보다 `includeSetupPage()`가 더 **이해하기 쉬운지** 이해가 안되네요.  

`함수명`+`인수` -> `함수명` 으로 단축되었기 때문일까요?

**Comment by sungyujeon** [comment, marked as answer]
1. includeSetupPage() 메서드는 `설정 페이지를 포함시켜라`라는 의미를 가지고, includeSetupPageInto(StringBuffer newPageContent) 메서드는 `새로운 페이지 내용에 설정 페이지를 포함시켜라`라는 의미로 볼 수 있을 것 같습니다. 단순하게만 보자면 설정 페이지를 포함시키는 행위를 실행하는 데 있어서 어떤 새로운 페이지 내용이 기반이 되는지에 대한 이해가 해당 메서드를 읽을 때 필요하게 되므로 더 이해하기 어렵다고 말한 것이 아닌가 싶습니다.(그냥 literally 정보가 하나 더 필요하니까 이해하기 어렵다는게 아닌가란..)

2. 실제 메서드를 살펴보면 결국 argument를 하나 더 받아야 한다는 문제도 있는게 아닌가란 생각이 듭니다. p.50에 보면 `StringBuffer를 살펴보자. 인스턴스 변수로 선언하는 대신 함수 인수로 넘기는 방법도 있었다. 하지만 그랬다면 코드를 읽는 사람은 StringBuffer를 발견할 때마다 의미를 해석해야 한다` 라고 한 것처럼 StringBuffer를 인스턴스 변수로 빼느냐 마느냐의 차이가 아닌가 싶습니다. 

   ```java
   private void includeSetupPage() {
     include("Setup", "-setup");
   }
   
   private void include(String pageName, String arg) {
     WikiPage inheritedPage = findInheritedPage(pageName);
     if (inheritedPage != null) {
       String pagePathName = getPathNameForPage(inheritedPage);
       buildIncludeDirective(pagePathName, arg);
     }
   }
   
   private void buildIncludeDirective(String pagePathName, String arg) {
     newPageContent
       .append("\n!include ")
       .append(arg)
       .append(" .")
       .append(pagePathName)
       .append("\n");
   }
   ```

   만약 StringBuffer newPageContent를 인스턴스 변수로 선언하지 않았다면 아래와 같은 코드가 되지 않을까 싶습니다.

   ```java
   private void includeSetupPageInto(StringBuffer newPageContent) {
     include("Setup", "-setup", newPageContent);
   }
   
   private void includeInto(StringBuffer newPageContent, String pageName, String arg) {
     WikiPage inheritedPage = findInheritedPage(pageName);
     if (inheritedPage != null) {
       String pagePathName = getPathNameForPage(inheritedPage);
       buildIncludeDirective(newPageContent, pagePathName, arg);
     }
   }
   
   private void buildIncludeDirectiveInto(StringBuffer newPageContent, String pagePathName, String arg) {
     newPageContent
       .append("\n!include ")
       .append(arg)
       .append(" .")
       .append(pagePathName)
       .append("\n");
   }
   ```

   결국 includeSetupPage에서는 `Setup page를 include 하라`는 행위 하나에 초점이 맞춰져야 하고, include 메서드에서는 `'Setup'이라는 pageName으로 찾아낸 WikiPage가 있다면 그 페이지의 pathName을 build하라`는 행위 하나에 초점이 맞춰져야 하지만, 인수로 newPageContent라는 StringBuffer를 인수로 받았기 때문에 '어디'라는, include 단계에서 굳이 알고 있지 않아도 되는 정보(build 할 때만 알면 되니까)까지 들고 있어야 해서 명료함이 떨어지지 않나 생각됩니다. 

   또 includePageContent(StringBuffer newPageContent)에도 argument가 추가되어야 하는 것 같습니다.
